File: sonnix_uploader.js
/**
 * BULK SONG UPLOAD SCRIPT FOR SONNIX
 * 
 * ✅ Firebase credentials are already configured!
 * 
 * HOW TO USE:
 * 1. Add your songs to the 'songs' array below (starting line 32)
 * 2. Run: node bulk-upload-songs.example.js
 * 
 * That's it! Your songs will appear at: https://sonnix.netlify.app/all-songs
 * 
 * REQUIREMENTS:
 * - Node.js installed (download from nodejs.org)
 */

// ============================================
// FIREBASE CONFIGURATION
// ============================================
// Firebase configuration for Sonnix
const firebaseConfig = {
  apiKey: "AIzaSyOgh68Cuqhxzm11VGVRcc2W4BYFXP4ZNOk",
  authDomain: "music-x-dfd87.firebaseapp.com",
  projectId: "music-x-dfd87",
  storageBucket: "music-x-dfd87.firebasestorage.app",
  messagingSenderId: "600929755806",
  appId: "1:600929755806:web:3e11645bd94118854618f"
};

// ============================================
// SONGS DATA - ADD YOUR SONGS HERE
// ============================================
const songs = [
  {
    title: "Midnight Rain",
    artist: "Taylor Swift",
    duration: 178,  // in seconds (2:58)
    audioUrl: "https://example.com/midnight-rain.mp3"
  },
  {
    title: "As It Was",
    artist: "Harry Styles",
    duration: 167,  // 2:47
    audioUrl: "https://example.com/as-it-was.mp3"
  },
  {
    title: "Anti-Hero",
    artist: "Taylor Swift",
    duration: 200,  // 3:20
    audioUrl: "https://example.com/anti-hero.mp3"
  }
  // Add more songs here...
];

// ============================================
// SCRIPT LOGIC (DO NOT EDIT BELOW)
// ============================================

import https from 'https';
import http from 'http';

// Simple Firebase REST API implementation
class FirebaseUploader {
  constructor(config) {
    this.projectId = config.projectId;
    this.apiKey = config.apiKey;
    this.baseUrl = `https://firestore.googleapis.com/v1/projects/${this.projectId}/databases/(default)/documents`;
  }

  async uploadSongs(songs) {
    console.log('\n🎵 Starting bulk song upload...\n');
    console.log(`Total songs to upload: ${songs.length}\n`);

    const results = {
      success: 0,
      failed: 0,
      errors: []
    };

    for (let i = 0; i < songs.length; i++) {
      const song = songs[i];
      console.log(`[${i + 1}/${songs.length}] Uploading: "${song.title}" by ${song.artist}...`);

      try {
        await this.uploadSong(song);
        results.success++;
        console.log(`✅ Success: "${song.title}"\n`);
      } catch (error) {
        results.failed++;
        results.errors.push({ song: song.title, error: error.message });
        console.error(`❌ Failed: "${song.title}" - ${error.message}\n`);
      }

      // Small delay to avoid rate limiting
      await this.sleep(500);
    }

    return results;
  }

  async uploadSong(song) {
    const songData = {
      fields: {
        title: { stringValue: song.title },
        artist: { stringValue: song.artist },
        duration: { integerValue: song.duration.toString() },
        audioUrl: { stringValue: song.audioUrl },
        album: { stringValue: song.album || '' },
        genre: { stringValue: song.genre || 'Unknown' },
        releaseDate: { stringValue: song.releaseDate || new Date().toISOString().split('T')[0] },
        imageUrl: { stringValue: song.imageUrl || '/default-song.png' },
        plays: { integerValue: '0' },
        isLiked: { booleanValue: false },
        createdAt: { timestampValue: new Date().toISOString() },
        updatedAt: { timestampValue: new Date().toISOString() },
        customId: { stringValue: `song_${Date.now()}_${Math.random().toString(36).substr(2, 9)}` }
      }
    };

    const url = `${this.baseUrl}/songs?key=${this.apiKey}`;
    
    return new Promise((resolve, reject) => {
      const urlObj = new URL(url);
      const options = {
        hostname: urlObj.hostname,
        path: urlObj.pathname + urlObj.search,
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      };

      const req = https.request(options, (res) => {
        let data = '';

        res.on('data', (chunk) => {
          data += chunk;
        });

        res.on('end', () => {
          if (res.statusCode === 200 || res.statusCode === 201) {
            resolve(JSON.parse(data));
          } else {
            reject(new Error(`HTTP ${res.statusCode}: ${data}`));
          }
        });
      });

      req.on('error', (error) => {
        reject(error);
      });

      req.write(JSON.stringify(songData));
      req.end();
    });
  }

  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

// Validate configuration
function validateConfig() {
  // Configuration is pre-filled for Sonnix project
  console.log('✅ Firebase configuration validated');
  console.log(`📦 Project: ${firebaseConfig.projectId}\n`);
}

// Validate songs data
function validateSongs() {
  if (!songs || songs.length === 0) {
    console.error('\n❌ ERROR: No songs found to upload!');
    console.error('\n📝 Please add songs to the "songs" array in this file.');
    console.error('\n');
    process.exit(1);
  }

  // Validate each song has required fields
  for (let i = 0; i < songs.length; i++) {
    const song = songs[i];
    if (!song.title || !song.artist || !song.duration || !song.audioUrl) {
      console.error(`\n❌ ERROR: Song at index ${i} is missing required fields!`);
      console.error('Required fields: title, artist, duration, audioUrl');
      console.error('Song data:', JSON.stringify(song, null, 2));
      console.error('\n');
      process.exit(1);
    }
  }
}

// Main execution
async function main() {
  console.clear();
  console.log('╔════════════════════════════════════════════╗');
  console.log('║   SONNIX BULK SONG UPLOAD TOOL            ║');
  console.log('║   Upload multiple songs programmatically  ║');
  console.log('╚════════════════════════════════════════════╝');
  console.log('');

  // Validate configuration and data
  validateConfig();
  validateSongs();

  // Create uploader and start upload
  const uploader = new FirebaseUploader(firebaseConfig);
  
  try {
    const results = await uploader.uploadSongs(songs);

    console.log('\n╔════════════════════════════════════════════╗');
    console.log('║           UPLOAD COMPLETE!                 ║');
    console.log('╚════════════════════════════════════════════╝');
    console.log(`\n✅ Successfully uploaded: ${results.success} songs`);
    console.log(`❌ Failed uploads: ${results.failed} songs`);

    if (results.errors.length > 0) {
      console.log('\n⚠️  Errors:');
      results.errors.forEach((err, index) => {
        console.log(`${index + 1}. ${err.song}: ${err.error}`);
      });
    }

    console.log('\n🎉 Your songs are now available at:');
    console.log('   https://sonnix.netlify.app/all-songs');
    console.log('\n');
  } catch (error) {
    console.error('\n❌ Fatal error:', error.message);
    process.exit(1);
  }
}

// Run the script
main();

File: sonnix_uploader.py
"""
Sonnix Firebase Uploader
Uploads songs to Sonnix website (Firebase) after successful Supabase upload
"""

import requests
import json
from datetime import datetime
from pathlib import Path
from mutagen.mp3 import MP3
from song_parser import parse_songs_from_ai_response
import re


# Firebase Configuration for Sonnix
FIREBASE_CONFIG = {
    "apiKey": "AIzaSyOgh68Cuqhxzm11VGVRcc2W4BYFXP4ZNOk",
    "projectId": "music-x-dfd87",
    "baseUrl": "https://firestore.googleapis.com/v1/projects/music-x-dfd87/databases/(default)/documents"
}


def extract_artist_and_song_from_filename(filename: str) -> dict:
    """
    Extract artist and song name from filename
    Expected formats:
    - "Song Name by Artist Name.mp3"
    - "Artist Name - Song Name.mp3"
    - "Song Name.mp3" (artist = Unknown)
    
    Returns: {'song': str, 'artist': str}
    """
    # Remove .mp3 extension
    name = filename.replace('.mp3', '').strip()
    
    # Pattern 1: "Song by Artist"
    match = re.match(r'^(.+?)\s+by\s+(.+?)$', name, re.IGNORECASE)
    if match:
        return {
            'song': match.group(1).strip(),
            'artist': match.group(2).strip()
        }
    
    # Pattern 2: "Artist - Song"
    if ' - ' in name:
        parts = name.split(' - ', 1)
        # Try to guess which is artist and which is song
        # Usually artist comes first in downloads
        return {
            'artist': parts[0].strip(),
            'song': parts[1].strip()
        }
    
    # Pattern 3: Just song name (no artist)
    return {
        'song': name,
        'artist': 'Unknown Artist'
    }


def get_audio_duration(file_path: str) -> int:
    """
    Get duration of MP3 file in seconds
    
    Args:
        file_path: Path to MP3 file
        
    Returns:
        Duration in seconds (integer)
    """
    try:
        audio = MP3(file_path)
        return int(audio.info.length)
    except Exception as e:
        print(f"⚠️ Could not read duration for {file_path}: {e}")
        return 180  # Default 3 minutes


def upload_song_to_sonnix(song_data: dict) -> dict:
    """
    Upload a single song to Sonnix Firebase
    
    Args:
        song_data: Dict with keys: title, artist, duration, audioUrl
        
    Returns:
        {'success': bool, 'message': str, 'firebase_id': str}
    """
    try:
        # Prepare Firebase document structure
        firebase_document = {
            "fields": {
                "title": {"stringValue": song_data['title']},
                "artist": {"stringValue": song_data['artist']},
                "duration": {"integerValue": str(song_data['duration'])},
                "audioUrl": {"stringValue": song_data['audioUrl']},
                "album": {"stringValue": song_data.get('album', '')},
                "genre": {"stringValue": song_data.get('genre', 'Unknown')},
                "releaseDate": {"stringValue": song_data.get('releaseDate', datetime.now().strftime('%Y-%m-%d'))},
                "imageUrl": {"stringValue": song_data.get('imageUrl', '/default-song.png')},
                "plays": {"integerValue": "0"},
                "isLiked": {"booleanValue": False},
                "createdAt": {"timestampValue": datetime.now().isoformat()},
                "updatedAt": {"timestampValue": datetime.now().isoformat()},
                "customId": {"stringValue": f"song_{int(datetime.now().timestamp())}_{song_data['title'][:10].replace(' ', '_')}"}
            }
        }
        
        # Firebase REST API endpoint
        url = f"{FIREBASE_CONFIG['baseUrl']}/songs?key={FIREBASE_CONFIG['apiKey']}"
        
        # Make POST request
        response = requests.post(
            url,
            headers={'Content-Type': 'application/json'},
            json=firebase_document,
            timeout=30
        )
        
        if response.status_code in [200, 201]:
            firebase_id = response.json().get('name', '').split('/')[-1]
            print(f"✅ Uploaded to Sonnix: \"{song_data['title']}\" by {song_data['artist']}")
            return {
                'success': True,
                'message': 'Uploaded to Sonnix successfully',
                'firebase_id': firebase_id
            }
        else:
            print(f"❌ Sonnix upload failed: HTTP {response.status_code}")
            print(f"Response: {response.text[:200]}")
            return {
                'success': False,
                'message': f'HTTP {response.status_code}: {response.text[:100]}'
            }
            
    except Exception as e:
        print(f"❌ Sonnix upload error: {str(e)}")
        return {
            'success': False,
            'message': str(e)
        }


def upload_batch_to_sonnix(audio_files: list, public_urls: list, progress_callback=None) -> dict:
    """
    Upload multiple audio files to Sonnix after Supabase upload
    
    Args:
        audio_files: List of local file paths
        public_urls: List of tuples [(filename, supabase_url), ...]
        progress_callback: Optional function to call with progress updates
        
    Returns:
        {'successful': int, 'failed': int, 'details': list}
    """
    def log(message):
        print(message)
        if progress_callback:
            progress_callback(message)
    
    log("🎸 Uploading to Sonnix...")
    
    results = {
        'successful': 0,
        'failed': 0,
        'details': []
    }
    
    # Create a mapping of filename to Supabase URL
    url_map = {filename: url for filename, url in public_urls}
    
    log(f"🎵 Uploading {len(audio_files)} songs to Sonnix Firebase...")
    
    for i, audio_file in enumerate(audio_files, 1):
        try:
            file_path = Path(audio_file)
            filename = file_path.name
            
            # Check if we have a Supabase URL for this file
            if filename not in url_map:
                log(f"⚠️ Skipping {filename}: No Supabase URL found")
                results['failed'] += 1
                results['details'].append({
                    'filename': filename,
                    'success': False,
                    'error': 'No Supabase URL'
                })
                continue
            
            supabase_url = url_map[filename]
            
            # Extract artist and song from filename
            metadata = extract_artist_and_song_from_filename(filename)
            
            # Get audio duration
            duration = get_audio_duration(str(file_path))
            
            # Prepare song data for Sonnix
            song_data = {
                'title': metadata['song'],
                'artist': metadata['artist'],
                'duration': duration,
                'audioUrl': supabase_url,
                'genre': 'Music',  # Default genre
                'album': '',
                'imageUrl': '/default-song.png'
            }
            
            # Upload to Sonnix
            log(f"📤 [{i}/{len(audio_files)}] {metadata['song']} by {metadata['artist']}")
            result = upload_song_to_sonnix(song_data)
            
            if result['success']:
                results['successful'] += 1
                log(f"✅ [{i}/{len(audio_files)}] Success")
            else:
                results['failed'] += 1
                log(f"❌ [{i}/{len(audio_files)}] Failed")
            
            results['details'].append({
                'filename': filename,
                'song': metadata['song'],
                'artist': metadata['artist'],
                'duration': duration,
                'supabase_url': supabase_url,
                'success': result['success'],
                'message': result.get('message', ''),
                'firebase_id': result.get('firebase_id', '')
            })
            
        except Exception as e:
            log(f"❌ Error processing {audio_file}: {str(e)}")
            results['failed'] += 1
            results['details'].append({
                'filename': Path(audio_file).name,
                'success': False,
                'error': str(e)
            })
    
    log(f"✅ Sonnix complete: {results['successful']}/{len(audio_files)} uploaded")
    if results['successful'] > 0:
        log(f"🎉 View at: https://sonnix.netlify.app/all-songs")
    
    return results


# Test function
if __name__ == "__main__":
    # Test artist/song extraction
    test_filenames = [
        "Shape of You by Ed Sheeran.mp3",
        "Taylor Swift - Anti-Hero.mp3",
        "Blinding Lights.mp3"
    ]
    
    print("Testing filename parsing:")
    for filename in test_filenames:
        result = extract_artist_and_song_from_filename(filename)
        print(f"  {filename}")
        print(f"    → Song: {result['song']}, Artist: {result['artist']}")
